### PR TITLE
feat: add SpikeSim event-driven energy estimator

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

Add a new `spikingjelly.activation_based.op_counter.spikesim` module for runtime event-driven energy estimation inspired by SpikeSim.

This keeps SpikeSim’s default energy coefficients and covered hardware components, but replaces the released dense `out_dim^2 * T` approximation with runtime event statistics from real `aten.convolution` calls. The result is energy estimation that is sensitive to actual spike activity while staying aligned with SpikeSim’s hardware accounting style.

## Changes

- Add `SpikeSimEnergyConfig`, `SpikeSimEventEnergyProfiler`, `SpikeSimEventEnergyReport`, and `estimate_spikesim_event_energy(...)`
- Implement runtime Conv2d interception with `TorchDispatchMode`
- Add `SpikeSimEventCounter` with:
  - `BaseCounter` integration
  - `aten` overload objects as rule keys
  - per-stage `A / R / Z` event statistics
  - proper `count / record / get_counts / get_total` behavior
- Limit v1 to Conv2d-stage energy only
- Warn or raise on unsupported ops/structures:
  - grouped/depthwise conv
  - transposed conv
  - linear-like ops
  - tensor-tensor merge ops such as residual adds / concat
- Export the new API from `op_counter`
- Add API docs for the new SpikeSim profiler

## Tests

Added coverage for:

- dense-equivalence on fully active binary inputs
- hand-checked sparse `A / R / Z` counts
- multi-step wrapper behavior
- mixed dense/event-driven stages
- repeated scope accumulation
- unsupported op warnings
- scalar add not triggering merge warnings
- tail-tile row cost handling
- config validation
- no-supported-stage warning
- shape-mismatch warning without crashing
- `BaseCounter` contract behavior

Local validation:

- `test/activation_based/test_op_counter_spikesim_event_energy.py` → `14 passed`
- `test/activation_based/test_synops.py test/activation_based/test_op_counter_neuromc_runtime.py test/activation_based/test_ac.py test/activation_based/test_mac.py` → `33 passed`

## Summary by CodeRabbit

* **New Features**
  * Added SpikeSim event-driven energy profiler with per-stage/per-component reports and a one-call helper.
  * Added SpikeSim energy configuration dataclass with validation, cloning, and computed energy accessors.
  * Re-exported SpikeSim public API at package level for easier access.

* **Documentation**
  * Added API docs for the SpikeSim energy profiling module.

* **Tests**
  * Added comprehensive tests for dense/sparse/mixed estimates, warnings/errors, context cleanup, and config validation.

### CODE REVIEW

**Comment by AllenYolk** [comment]
Addressed the latest review feedback:\n\n- aligned SpikeSim docs/export surface and public API docstrings\n- hardened profiler context cleanup on enter/exit\n- switched runtime occupancy counting to float32\n- added fail-fast validation for tile-length mismatches in formulas\n- expanded local tests and reran SpikeSim/op_counter regressions\n\n@gemini-code-assist[bot] please re-review the updated implementation, with extra attention to correctness of the event-driven energy math and consistency with the existing op_counter/neuromc style.

**Comment by AllenYolk** [comment]
Addressed the remaining review nitpicks:\n\n- broadened  input typing to match sequence-valued stats\n- converted  to a plain dict before returning the report\n\nLocal validation rerun after the patch:\n-  -> 15 passed\n-  -> 33 passed

**Comment by AllenYolk** [comment]
Addressed the remaining review nitpicks:

- broadened `spikingjelly.activation_based.op_counter.spikesim.formulas` input typing to match sequence-valued stats
- converted `energy_by_component["totals"]` to a plain `dict` before returning the report

Local validation rerun after the patch:
- `test_op_counter_spikesim_event_energy.py` -> 15 passed
- `test_synops.py + test_op_counter_neuromc_runtime.py + test_ac.py + test_mac.py` -> 33 passed

**Comment by AllenYolk** [comment]
@gemini-code-assist  review\

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+        self.rules = {
+            aten.convolution.default: self._count_convolution,
+            aten.add.Tensor: self._count_merge,
+            aten.add_.Tensor: self._count_merge,
+            aten.mm.default: self._count_linear_like,
+            aten.addmm.default: self._count_linear_like,
+            aten.bmm.default: self._count_linear_like,
+            aten.baddbmm.default: self._count_linear_like,
+            aten.cat.default: self._count_concat,
+        }
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply spikingjelly/activation_based/op_counter/spikesim/counter.py:105]: `SpikeSimEventCounter` is a subclass of `BaseCounter`. `has_rule` is defined in `BaseCounter`. @gemini-code-assist

**Comment by AllenYolk** [comment]
Addressed the latest bot feedback from the current review loop:

- added a strict "both-or-neither" check for `active_row_count_by_tile` and `input_tile_channels`
- added a targeted formula-level test for that fail-fast path

Revalidated locally:
- `test_op_counter_spikesim_event_energy.py` -> 16 passed
- `test_synops.py + test_op_counter_neuromc_runtime.py + test_ac.py + test_mac.py` -> 33 passed

I also re-checked the `has_rule` concern: `SpikeSimEventCounter` inherits `BaseCounter`, so `has_rule()` is already present through the base class and is working with the current `aten`-overload keyed rules.

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. Please focus on any remaining correctness or style issues in the SpikeSim event-driven energy estimator and its tests.

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. I addressed the current formula validation path and explicitly added `has_rule()` on `SpikeSimEventCounter` for clarity, then reran local targeted/regression tests.

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. I addressed the current performance-oriented feedback in SpikeSim event counting and reran local targeted/regression tests.

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. I addressed the current performance-oriented feedback in SpikeSim event counting and reran local targeted/regression tests.

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. I addressed the current code-style/memory-efficiency suggestions in SpikeSim event counting and reran local targeted/regression tests.

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. I addressed the current cleanup/performance comments around SpikeSim tensor handling and xbar-row energy documentation, then reran local targeted/regression tests.

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Comment** — Solid v1 implementation with clean architecture. A few items to address before merge, none critical.

---

### ✅ Looks Good

- Clean module separation: `config.py`, `counter.py`, `trace.py`, `formulas.py`, `core.py` — easy to navigate.
- `BaseCounter` integration is correct; `count()` properly dispatches to `record()` for both `"Global"` and scoped names.
- `_exclude_python_dispatch_guard()` correctly prevents re-entrant dispatch from `_spike_event_counts`'s internal `F.conv2d` call.
- Context manager cleanup in `__enter__`/`__exit__` is correct — if `_trace_mode.__enter__` fails, `_tracker.__exit__` is called in the exception handler.
- The warn-or-raise pattern (`_warn_or_raise`) with dedup via `_warning_keys` is clean.
- Test coverage is comprehensive: 14 tests covering dense/sparse equivalence, edge cases, config validation, context cleanup, and monotonicity.
- Docstrings are bilingual and follow the project's RST conventions.

---

### ⚠️ Warnings

1. **`formulas.py:66` — Dead code: `_compute_spikesim_dense_stage_energy`**
   This function is defined but never called anywhere in the codebase (neither in production code nor tests). The test file uses its own `_dense_stage_energy` helper instead. Remove it or reconcile with the test helper to avoid drift.

2. **`config.py:22` — Chinese docstring section is in English**
   The "中文" section of `SpikeSimEnergyConfig`'s class docstring contains the same English text as the "English" section. Per the project's AGENTS.md, both sections must be present and semantically equivalent but in their respective languages. This needs a proper Chinese translation.

3. **`counter.py:210-213` — `_pair_tuple` doesn't guard against non-sequence inputs**
   If `x` is not a sequence (e.g., `None`, an int that's not caught by the first branch), `len(x)` will raise `TypeError` before the intended `ValueError`. Consider adding an `isinstance(x, (list, tuple))` check.

4. **`core.py:143` — `get_counts()` shadows `BaseCounter.get_counts()` with incompatible semantics**
   `BaseCounter.get_counts()` returns `dict[str, dict[Any, int]]` (op counts), but the profiler's `get_counts()` returns `dict[str, dict[str, Any]]` (event stats + metadata). Users who rely on the base class interface will get surprised. Consider renaming to `get_event_counts()` or documenting the divergence prominently.

---

### 💡 Suggestions

1. **`test_op_counter_spikesim_event_energy.py:326-327` — Flaky monotonicity test**
   `test_spikesim_event_energy_monotonic_with_input_sparsity` uses `torch.rand` without a fixed seed. With 80% vs 20% density thresholds, a sparse input could coincidentally have more active patches than a dense one in unlucky draws. Fix: use `torch.manual_seed` or construct deterministic inputs (e.g., `torch.ones` with explicit zero-ing).

2. **`counter.py:35` — `_is_spike_like` could be clearer**
   `not bool(((x != 0) & (x != 1)).any().item())` is correct but reads inverted. Consider:
   ```python
   return bool((x == 0) | (x == 1)).all().item()
   ```
   Or just use `x.is_floating_point()` + compare against `{0, 1}` for clarity.

3. **`config.py:9` — `_SPIKESIM_XBAR_ROW_DIVISOR = 8.0` lacks explanation**
   This constant is used in `xbar_row_energy_pj()` but the comment only says "Keep the row-level basis aligned with the released SpikeSim dense energy path." A one-liner explaining that this corresponds to 8 read ports per xbar row would help future maintainers.

4. **`core.py:115` — `get_report()` accumulates `component_totals["total_pj"]` alongside sub-components**
   The `totals_dict` will contain `"total_pj"` as a key alongside `"patch_control_pj"`, `"xbar_pj"`, `"neuron_pj"`. This is semantically correct but could confuse consumers who iterate `component_totals.items()` expecting only leaf components. Consider separating total into its own field or documenting the nesting.

5. **`SpikeSimEnergyConfig` — No `__post_init__` validation**
   Users can create `SpikeSimEnergyConfig(xbar_size=-1)` without error until the profiler's constructor calls `validate()`. Consider adding `__post_init__: self.validate()` for fail-fast, or document that validation is deferred.

---

### 🔴 Critical
None.

---

*Reviewed by Hermes Agent @ `0e7876d`*

File: spikingjelly/activation_based/op_counter/spikesim/formulas.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    return {
+        "patch_control_pj": patch_control,
+        "xbar_pj": xbar,
+        "neuron_pj": neuron,
+        "total_pj": patch_control + xbar + neuron,
+    }
+
+
+def _compute_spikesim_dense_stage_energy(
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/formulas.py:66]: ⚠️ **Dead code:** `_compute_spikesim_dense_stage_energy` is defined here but never called anywhere — neither in production code nor tests. The test file uses its own `_dense_stage_energy` helper instead. Remove or reconcile to avoid drift.

File: spikingjelly/activation_based/op_counter/spikesim/config.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+_SPIKESIM_XBAR_ROW_DIVISOR = 8.0
+
+
+@dataclass
+class SpikeSimEnergyConfig:
+    r"""
+    **API Language:**
+    :ref:`中文 <SpikeSimEnergyConfig-cn>` | :ref:`English <SpikeSimEnergyConfig-en>`
+
+    ----
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/config.py:16]: ⚠️ **Docstring issue:** The 中文 section here contains English text identical to the English section. Per AGENTS.md, both sections must be present and semantically equivalent but in their respective languages. This needs a proper Chinese translation.

```rst
* **中文**

Runtime energy configuration for the event-driven SpikeSim estimator.
```
should be something like:
```rst
* **中文**

SpikeSim 事件驱动能耗估计器的运行时能耗配置。
```

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+from ..base import BaseCounter
+from .config import SpikeSimEnergyConfig
+
+__all__ = []
+aten = torch.ops.aten
+
+
+def _shape_tuple(x: torch.Tensor) -> tuple[int, ...]:
+    return tuple(int(v) for v in x.shape)
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/counter.py:19]: 💡 **Minor robustness:** If `x` is not a sequence (e.g., a non-int non-sequence), `len(x)` will raise `TypeError` before the intended `ValueError`. Consider:
```python
def _pair_tuple(x: Any) -> tuple[int, int]:
    if isinstance(x, int):
        return (int(x), int(x))
    if not isinstance(x, (list, tuple)) or len(x) != 2:
        raise ValueError(f"expected a pair, but got {x}.")
    return (int(x[0]), int(x[1]))
```

File: spikingjelly/activation_based/op_counter/spikesim/core.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        self.strict = strict
+        self.verbose = verbose
+        self._counter = SpikeSimEventCounter(
+            config=self.config,
+            strict=self.strict,
+            verbose=self.verbose,
+        )
+        self._tracker = ActiveModuleTracker()
+        self._trace_mode = SpikeSimEventTraceMode(
+            self._tracker,
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/core.py:143]: 💡 **API divergence:** `BaseCounter.get_counts()` returns `dict[str, dict[Any, int]]` (op counts), but this `get_counts()` returns `dict[str, dict[str, Any]]` (event stats + metadata). Users relying on the base class interface will get surprised. Consider renaming to `get_event_counts()` or documenting the divergence prominently in the docstring.

File: test/activation_based/test_op_counter_spikesim_event_energy.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        model_s.weight.copy_(model_m.weight)
+
+    x_seq = (torch.rand(3, 2, 3, 8, 8) > 0.7).float()
+    report_m = op_counter.estimate_spikesim_event_energy(model_m, x_seq, config=config)
+
+    total_s = 0.0
+    for t in range(x_seq.shape[0]):
+        total_s += op_counter.estimate_spikesim_event_energy(
+            model_s, x_seq[t], config=config
+        ).energy_total_pj
```
Comment (AllenYolk) [inline test/activation_based/test_op_counter_spikesim_event_energy.py:118]: 💡 **Flaky test risk:** `torch.rand` without a fixed seed means this test could fail non-deterministically. With 80% vs 20% density thresholds, an unlucky draw could produce `sparse_x` with more active patches than `dense_x`. Fix: use `torch.manual_seed(42)` at the start, or construct deterministic inputs (e.g., `torch.ones(..., 1)` with explicit channel zeroing).

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. I addressed the latest precision-related feedback in SpikeSim occupancy counting and reran local targeted/regression tests.

**Review by AllenYolk — COMMENTED** [review]
## Code Review Summary

**Verdict: Approved ✅** (0 critical, 0 warnings, 4 suggestions)

**PR:** #673 — feat: add SpikeSim event-driven energy estimator
**Author:** @AllenYolk
**Files changed:** 11 (+1338 -4)

### 🔴 Critical

_None_

### ⚠️ Warnings

_None_

### 💡 Suggestions

- **spikesim/formulas.py:57** — `_compute_spikesim_dense_stage_energy` is defined but never called anywhere in production code or tests. The test file defines its own `_dense_stage_energy`. Consider removing to avoid dead code, or making it public if it's intended as a reference utility.

- **spikesim/counter.py:75** — `has_rule` is an exact copy of `BaseCounter.has_rule`. You can drop the override and inherit it. Same for the `count()` override — the base class already calls `self.rules[func](...)`, so the only added value is the two `self.record()` calls, which could live in a helper.

- **spikesim/config.py** — Chinese and English docstrings in `SpikeSimEnergyConfig` are character-for-character identical. The project convention asks for semantically consistent but distinct language versions. Consider adding Chinese-specific context (e.g., referencing the SpikeSim paper or the hardware model assumptions).

- **neuromc/core.py, neuromc/residency.py** — Formatting-only changes (line wrapping, trailing newlines) are mixed into this feature PR. They're harmless but would be cleaner as a separate commit to keep the diff focused.

### ✅ Looks Good

- Clean module decomposition: `config`, `counter`, `formulas`, `trace`, `core` — each has a single responsibility
- Correct BaseCounter integration: `count()` override properly records to both "Global" and per-scope, and `__torch_dispatch__` in `SpikeSimEventTraceMode` correctly routes only to `count()` (no double-recording)
- The event-driven counting logic in `_spike_event_counts` is well-designed: tile-aware channel summing → occupancy via conv2d with ones kernel → active patch/row/site counts. The `_exclude_python_dispatch_guard` correctly prevents re-dispatch of the internal conv2d call
- Graceful degradation on unsupported ops (grouped/depthwise conv, transposed conv, linear-like, concat, merge) — warns by default, raises in strict mode
- `_is_spike_like` correctly handles bool dtype fast-path and short-circuits via `.any()` for float tensors
- `_ones_kernel_cache` avoids repeated tensor allocation for the occupancy computation
- Thorough test coverage (14 tests): dense equivalence, sparse hand-checked counts, multi-step wrapper, mixed dense/event stages, repeated scope accumulation, unsupported op warnings, scalar add not triggering merge, tail-tile row cost, config validation, no-stage warning, shape mismatch, formula validation, context cleanup on enter failure, BaseCounter contract
- Bilingual docstrings follow project convention throughout
- Proper `__all__` exports at both module and package level

---
*Reviewed by Hermes Agent*

File: spikingjelly/activation_based/op_counter/spikesim/formulas.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    else:
+        xbar = (
+            float(metadata["out_channel_tiles"])
+            * float(stats["active_row_count"])
+            * config.xbar_row_energy_pj(metadata["in_channels"])
+        )
+    neuron = (
+        float(p_i) * float(stats["active_output_tile_site_count"]) * config.neuron_pj
+    )
+
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/formulas.py:57]: 💡 **Suggestion:** `_compute_spikesim_dense_stage_energy` is defined but never called in production code or tests (the test file defines its own `_dense_stage_energy`). Consider removing to avoid dead code, or making it public if intended as a reference utility.

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+    input_shape: tuple[int, ...]
+    output_shape: tuple[int, ...]
+    out_channel_tiles: int
+    input_tile_channels: list[int]
+    total_calls: int = 0
+    event_driven_calls: int = 0
+    dense_fallback_calls: int = 0
+    shape_mismatch_detected: bool = False
+
+    def as_dict(self) -> dict[str, Any]:
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/counter.py:75]: 💡 **Suggestion:** `has_rule` is an exact copy of `BaseCounter.has_rule` — you can drop the override and inherit it. Same for `count()` — the base class already calls `self.rules[func](...)`, so the only added value here is the two `self.record()` calls, which could be extracted to a small helper.

File: spikingjelly/activation_based/op_counter/spikesim/config.py (hunk 3 above)
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/config.py:16]: 💡 **Suggestion:** Chinese and English docstrings here are character-for-character identical. The project convention asks for semantically consistent but distinct language versions. Consider adding Chinese-specific context (e.g., referencing the SpikeSim paper or hardware model assumptions in Chinese).

File: spikingjelly/activation_based/op_counter/neuromc/core.py [hunk 9]
```diff
@@ -580,7 +580,9 @@ def get_report(self) -> NeuroMCRuntimeEnergyReport:
             for stage, op_info in self._stage_move_bits_by_op.items()
         }
 
-        rw_labeled_totals: dict[str, dict[str, int]] = defaultdict(lambda: defaultdict(int))
+        rw_labeled_totals: dict[str, dict[str, int]] = defaultdict(
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/neuromc/core.py:583]: 💡 **Suggestion:** The changes to `neuromc/core.py` and `neuromc/residency.py` are formatting-only (line wrapping, trailing newlines). They are harmless but would be cleaner as a separate commit to keep this PR diff focused on the SpikeSim feature.

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. I addressed the current config-safety/docstring feedback in SpikeSim config and reran local targeted/regression tests.

**Review by AllenYolk — COMMENTED** [review]
## Code Review Summary

**Verdict: APPROVE** — clean implementation, solid test coverage, no blocking issues.

Reviewed: `spikesim/` module (6 new files), `op_counter/__init__.py`, docs, 2 minor formatting changes in `neuromc/`, and 18 new tests.

**Local validation:** 18 new tests passed, 33 existing tests unaffected.

---

### ✅ Looks Good

- Clean module decomposition (`config`, `core`, `counter`, `formulas`, `trace`) with clear separation of concerns
- Consistent bilingual docstrings following project conventions
- Context manager cleanup on `__enter__` failure is correct and tested
- Warning deduplication via `_warning_keys` prevents spam
- `_exclude_python_dispatch_guard()` correctly prevents infinite recursion in the occupancy computation
- `aten.add.Scalar` is not in `rules`, so scalar adds silently pass without false warnings — correct by design
- `_ones_kernel_cache` is bounded by unique `(num_tiles, k_h, k_w, device)` combos — practical for profiling
- Dense fallback for non-spike inputs is a sensible v1 limitation
- `BaseCounter` contract is satisfied (records, get_counts, get_total)

### 💡 Suggestions (non-blocking)

1. **`formulas.py:53-83` — Dead code:** `_compute_spikesim_dense_stage_energy` is defined but never imported or called anywhere. Either remove it or add a test + use it as a reference implementation for dense-equivalence validation.

2. **`counter.py:30` — `_is_spike_like` full-tensor scan:** `((x != 0) & (x != 1)).any().item()` scans the entire input tensor. For v1 this is fine, but worth noting: if profiling large-batch inputs, this becomes a bottleneck. A future optimization could short-circuit on a random sample.

3. **`trace.py:5` — Private PyTAPI coupling:** `torch._C._ExcludeDispatchKeyGuard` and `torch._C.DispatchKeySet` are internal APIs. A brief comment explaining *why* this is needed (prevent dispatch recursion during occupancy computation) would help future maintainers.

4. **`neuromc/core.py`, `neuromc/residency.py` — Cosmetic changes in feature PR:** The line-wrapping and blank-line additions are fine individually but add noise to a 1349-line feature diff. Consider separating formatting into its own commit for cleaner git history.

5. **`counter.py:89` — Redundant `has_rule` override:** `SpikeSimEventCounter.has_rule` does exactly `return func in self.rules`, identical to `BaseCounter.has_rule`. The override is harmless but unnecessary.

6. **Test coverage gaps (minor):**
   - No test for `sram` device type (all tests use default `rram`)
   - No test for `verbose` mode output
   - No test for `SpikeSimEnergyConfig.copy()`
   - `test_spikesim_event_energy_counter_base_contract_nonempty` accesses `_counter` (private) — consider testing via the public `get_counts()` on the profiler instead

---

*Reviewed by Hermes Agent*

File: spikingjelly/activation_based/op_counter/spikesim/formulas.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            xbar_rows += float(row_count) * config.xbar_row_energy_pj(
+                int(tile_channels)
+            )
+        xbar = float(metadata["out_channel_tiles"]) * xbar_rows
+    else:
+        xbar = (
+            float(metadata["out_channel_tiles"])
+            * float(stats["active_row_count"])
+            * config.xbar_row_energy_pj(metadata["in_channels"])
+        )
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/formulas.py:53]: 💡 `_compute_spikesim_dense_stage_energy` is defined but never imported or called anywhere in the codebase. Dead code — either remove it or add a test that uses it as a reference for dense-equivalence validation.

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+
+def _shape_tuple(x: torch.Tensor) -> tuple[int, ...]:
+    return tuple(int(v) for v in x.shape)
+
+
+def _pair_tuple(x: Any) -> tuple[int, int]:
+    if isinstance(x, int):
+        return (int(x), int(x))
+    if len(x) != 2:
+        raise ValueError(f"expected a pair, but got {x}.")
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/counter.py:26]: 💡 `torch._C._ExcludeDispatchKeyGuard` and `torch._C.DispatchKeySet` are private PyTorch APIs. A brief comment explaining *why* this guard is needed (prevent dispatch recursion during the occupancy `F.conv2d` call) would help future maintainers when PyTorch internals change.

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+    def __init__(
+        self,
+        *,
+        config: SpikeSimEnergyConfig,
+        strict: bool,
+        verbose: bool,
+    ):
+        super().__init__()
+        self.config = config
+        self.strict = strict
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/counter.py:89]: 💡 This override is identical to `BaseCounter.has_rule` (`return func in self.rules`). Harmless but redundant — removing it would reduce surface area.

**Comment by AllenYolk** [comment]
@coderabbitai review the latest changes.
@gemini-code-assist[bot] review the latest changes. I addressed the latest docstring/comma and `reshape()` feedback in SpikeSim and reran local targeted/regression tests.

**Review by AllenYolk — COMMENTED** [review]
## Code Review Summary

**Verdict: APPROVE** — no critical or blocking issues found.

### ✅ Looks Good

- **Clean module structure**: `config.py`, `counter.py`, `trace.py`, `formulas.py`, `core.py` — each has a clear single responsibility.
- **Correct SpikeSim energy alignment**: Default coefficients and formulas match the published `ela_spikesim.py` dense energy path. The event-driven replacement is well-motivated.
- **Proper BaseCounter integration**: `SpikeSimEventCounter` correctly implements `count()`, `has_rule()`, and uses `aten` overload objects as rule keys.
- **Robust context manager**: `SpikeSimEventEnergyProfiler.__enter__` properly cleans up the tracker on trace_mode failure. `__exit__` chains both cleanup paths.
- **Config immutability**: Config is always `.copy()`'d before use — no mutation of user-provided instances.
- **Good warning/strict duality**: `_warn_or_raise` with dedup keys prevents spam; `strict=True` raises `NotImplementedError` immediately.
- **`_exclude_python_dispatch_guard`**: Correctly prevents recursive dispatch when the counter calls `F.conv2d` internally.
- **`_ones_kernel_cache`**: Good caching pattern to avoid repeated kernel allocation.
- **Bilingual docstrings**: All public APIs have Chinese + English docstrings with RST-style fields.
- **Test coverage**: 18 tests covering dense equivalence, sparse counts, mixed stages, multi-step, warnings, config validation, context cleanup, and monotonicity. All pass. Existing 33 tests unaffected.

### 💡 Suggestions (non-blocking)

1. **`_is_spike_like` scans the full tensor** (`counter.py:30-33`): For large activations (e.g., 256×56×56), `((x != 0) & (x != 1)).any().item()` is an O(n) pass just to determine event vs. dense mode. Consider sampling a small random subset (e.g., 64 elements) for a probabilistic fast-path, with a comment explaining the tradeoff. Not blocking for v1 — correctness is more important here.

2. **Scalar tensor edge case in `_count_merge`** (`counter.py:363`): `torch.is_tensor(args[1])` would return `True` for `tensor + torch.tensor(1.5)`, producing a false-positive merge warning. This is an unusual pattern in SNN models, but if you want to be precise, you could add a check like `args[1].numel() == 1` to treat single-element tensors as scalars.

3. **`_update_stage` tile accumulation is dense** (`counter.py:282-304`): The per-tile accumulation with `None` fallback and length mismatch handling is correct but reads heavily. Extracting a small `_accumulate_by_tile(existing, new)` helper would improve readability.

4. **Missing test for `get_counts()` and `get_total()`** on the profiler: `test_spikesim_event_energy_counter_base_contract_nonempty` tests `_counter.get_counts()`, but the public `profiler.get_counts()` and `profiler.get_total()` methods are not directly tested. Minor — they're thin wrappers.

5. **SRAM device mode untested**: All tests use the default `device="rram"`. A single test with `device="sram"` would verify the `xbar_array_energy_pj` property path.

---
*Reviewed by Hermes Agent — all 18 new + 33 existing tests pass locally.*

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+        return (int(x), int(x))
+    if len(x) != 2:
+        raise ValueError(f"expected a pair, but got {x}.")
+    return (int(x[0]), int(x[1]))
+
+
+def _is_spike_like(x: torch.Tensor) -> bool:
+    if x.dtype == torch.bool:
+        return True
+    return not bool(((x != 0) & (x != 1)).any().item())
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/counter.py:33]: 💡 **Suggestion:** For large activations (e.g., 256×56×56), this full-tensor scan is O(n) just to pick event vs. dense mode. Consider sampling a small random subset (e.g., 64 elements) as a probabilistic fast-path. Not blocking for v1 — correctness first.

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+            )
+
+    def _count_convolution(
+        self, args: tuple[Any, ...], kwargs: dict[str, Any], out
+    ) -> int:
+        self._handle_convolution(self._scope_from_kwargs(kwargs), args, out)
+        return 1
+
+    def _count_merge(self, args: tuple[Any, ...], kwargs: dict[str, Any], out) -> int:
+        if len(args) < 2 or (not torch.is_tensor(args[1])):
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/counter.py:363]: 💡 **Suggestion:** `torch.is_tensor(args[1])` would return `True` for `tensor + torch.tensor(1.5)`, producing a false-positive merge warning. Adding `and args[1].numel() != 1` would handle single-element tensors as scalars. Minor edge case — unlikely in real SNN models.

**Comment by AllenYolk** [comment]
Only maintenance-related comments are raised. No need for further revision.
